api/envoy/api/v2/cds.proto

```diff
@@ -234,6 +235,41 @@ message Cluster {
   //   verification.
   auth.UpstreamTlsContext tls_context = 11;
 
+  // TransportSocketMatch specifies what transport socket config will be used
+  // when the matches conditions are satisfied.
+  message TransportSocketMatcher {
+    // The name of the match.
+    string name = 1;
+    // In order to be match, an endpoint must have all the label key/value pairs
+    // as specified in by *label_match*.
+    repeated endpoint.Label label_matches = 2;
+    // The configuration of the transport socket.
+    core.TransportSocket transport_socket = 3;
+  }
+
+  // Configuration to use different transport sockets for different endpoints.
+  // This should only be used for EDS typed cluster.
+  //
+  // An endpoint's *endpoint.Labels* is used to match against the *label_matches*
+  // to decide which transport socket to use when connecting to that particular endpoint.
+  // For example, with following matcher
+  //
+  // .. code-block:: json
+  //
+  // {
+  //   "endpoint_transport_socket_matchers": [
+  //     { "name": "enable-mtls", label_matches: [ { "label": "app": value: "foo" } ]
+  //     transport_socket: { "name": "tls", "config": { ... }
+  //   ]
+  // }
+  //
+  // Connection to the endpoints whose "app" label equals to "foo" value uses "enable-mtls"
+  // configuration, a tls socket with detailed configuration specified as *config*.
+  //
+  // If an endpoint can't be matched against any *TransportSockerMatcher*, connection socket
+  // fallbacks to use the *tls_context* or *transport_socket* specified in this cluster.
+  repeated TransportSocketMatcher endpoint_transport_socket_matchers = 41;
```

Comment:
add not-implemented hide annotation, and add a TODO to deprecate transport_socket.

Reply:
By deprecation, do you envision there'll be a default transport socket configuration in the list, in order to replace `fallback to outer transport socket/tls_context` field?

I admit that'll be cleaner, just want to confirm.

Reply:
With more thoughts, I think we have to use outer defined `Cluster.transport_socket` can't deprecate it.

This is because for other typed discovery cluster (STRICT_DNS), there's no clear semantics how can they be incorporated into this added fields. It applies only to EDS cluster.

Reply:
`not-implemented` added.

Reply:
> With more thoughts, I think we have to use outer defined Cluster.transport_socket can't deprecate it.
> This is because for other typed discovery cluster (STRICT_DNS), there's no clear semantics how can they be incorporated into this added fields. It applies only to EDS cluster.

That's another reason I don't like @rshriram's string to transport socket map because it doesn't have default semantics, and much less flexibility for other types.

@rshriram @htuch can we go back to the doc and discuss more, I don't like the string map for the reason above and below.

Reply:
I add a section in the [doc](https://docs.google.com/document/d/1lpshO7E-2KQ5SC519oiRe25TPAXp1HUcyVzZ4QM4IeU/edit?userstoinvite=vladimir%40spotify.com&ts=5d6679a9&actionButton=1#heading=h.3ffxyg8crnt3), and trying to list the pros and cons, PTAL, thanks.

Reply:
Sure, let's take this to the doc.